### PR TITLE
[SPARK-31578][R] Vectorize schema validation for arrow in types.R

### PR DESCRIPTION
**Pull request by MichaelChirico**
### What changes were proposed in this pull request?

Repeated `sapply` avoided in internal `checkSchemaInArrow`

### Why are the changes needed?

Current implementation is doubly inefficient:

 1. Repeatedly doing the same (95%) `sapply` loop
 2. Doing scalar `==` on a vector (`==` should be done over the whole vector for efficiency)

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

By my trusty friend the CI bots

### CODE REVIEW

File: R/pkg/R/types.R [hunk 1]
```diff
@@ -94,27 +94,23 @@ checkSchemaInArrow <- function(schema) {
...
+  if (any(startsWith(field_strings, "StructType"))) {
     stop("Arrow optimization in R does not support nested struct type yet.")
   }
-  if (any(sapply(schema$fields(),
-                 function(x) startsWith(x$dataType.toString(),
-                 "MapType")))) {
+  if (any(startsWith(field_strings, "MapType"))) {
     stop("Arrow optimization in R does not support map type yet.")
   }
+  return(invisible())
```
Comment (HyukjinKwon) [inline R/pkg/R/types.R:115]: Quick question, why do we need this?

**Review by HyukjinKwon — APPROVED** [review]

File: R/pkg/R/types.R (hunk 1 above)
Comment (MichaelChirico) [reply R/pkg/R/types.R:115]: Just to be explicit about a void return value. Not strictly necessary. Happy to remove.

Comment (HyukjinKwon) [reply R/pkg/R/types.R:115]: I think this is fine. Let's just remove unless this is explicitly recommended, conventional or consistent across the codebase :-).

**Comment by HyukjinKwon** [comment]
I am going to merge. The tests are already passed, and I don't believe removing the return causes any test or build to break.

**Comment by MichaelChirico** [comment]
linter failure looks transient to me:

> E: The repository 'https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/ Release' does not have a Release file.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Pull request closed without merging**